docs/e2e-testing-headscale-derper-tailscale.md

```diff
@@ -0,0 +1,368 @@
+# End to end testing: Tailscale, Headscale, Derper
+
+
+## Deploy environment
+
+For this testing, we'll use VMs on Azure.
+This ensures we can test scenarios that are close to real world,
+with public IP addresses and DNS names for Headscale and Derper.
+
+This document deploys everything in the `australiaeast` region;
+if you deploy in another region, substitute the name accordingly in domain names where referenced.
+
+### Terraform
+
+There is a terraform module at [./e2e.tf](./e2e.tf) that will deploy the environment.
+
+There are multiple ways to authenticate, but one way is to ensure the `az` command (azure cli) is installed and authenticated.
+See https://learn.microsoft.com/en-gb/cli/azure/get-started-with-azure-cli for more information.
+
+Then you will need `ARM_SUBSCRIPTION_ID` in the environment - see:
+https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/azure_cli for details.
+Example:
+
+```bash
+export ARM_SUBSCRIPTION_ID=466ee356-ce1f-11ef-bf94-db2a042e145c
+```
+
+Then you can apply the terraform, providing required variables:
+
+```bash
+terraform apply -var "ssh_key=YOUR PUBLIC SSH KEY CONTENTS"
+```
+
+### Manually
+
+We'll deploy 5 VMs:
+
+| name (same DNS name) | full domain name                             | inbound rules                                                                                           | outbound rules |
+|----------------------|----------------------------------------------|---------------------------------------------------------------------------------------------------------|----------------|
+| headscale            | headscale.australiaeast.cloudapp.azure.com   | - TCP 22 (ssh)<br>- TCP 80 (letsencrypt challenge)<br>- TCP 443 (headscale)                             | None           |
+| derper               | derper.australiaeast.cloudapp.azure.com      | - TCP 22 (ssh)<br>- TCP 80 (derper)<br>- TCP 443 (derper)<br>- UDP 3478 (derper STUN)<br>- ICMP traffic | - ICMP traffic |
+| tailscale-1          | tailscale-1.australiaeast.cloudapp.azure.com | - TCP 22 (ssh)                                                                                          | None           |
```

Comment:
For typical use cases, I think the public IP/domain name should not be available to the client side. Clients are often behind NAT and use private IPs. Furthermore, is it possible to put tailscale-1 / tailscale-2 / tailscale-3 into different regions?

Reply:
Agreed, but I need access so I can ssh in to install and test tailscale. Not sure of an alternative - perhaps I can set up a jumpbox situation ([machine with public ip] -> [tailscale machine in same internal network, no public ip]) for testing?

I'll see what I can do.

Reply:
I've set up jumpboxes to two machines on separate private networks, so I think that covers it. :)